### DISCUSSION TITLE
Breaking changes post-v1

### DISCUSSION THREAD

**Discussion by mansellan** [twinbasic/twinbasic#1452] [General]
In a greenfield project, once a project reaches v1 it's best to try and avoid breaking changes, almost at any cost. However, twinBASIC is not strictly greenfield, given that it has an absolute design goal to be 100% backwards-compatible with source code written for legacy VB6 SP6/ VBA6 / VBA7 (aka VBx).

VBx is not fully specified. There are API docs, but VBx has lots of undocumented implementation details (quirks) that have to be accommodated if source code is to be fully portable from VBx to twinBASIC. The current level of compatibility is impressive, but edge-cases continue to be unearthed. Given the opaque nature of VBx, this is likely to be the case for some time.

This raises an interesting dilemma. Once twinBASIC v1 has been released, what should be done when back-compat issues are discovered? Especially when a fix could create breaking changes between twinBASIC versions?

Not sure what the answer is, would like to hear opinions.

----

Postscript:

My opinion is that backwards-compatibility should usually (always?) trump concerns of breaking changes between twinBASIC versions, as there is an enormous amount of code in VBx and (currently!) very little targeting twinBASIC specifically. Any breaking changes in twinBASIC post-v1 should of course be flagged in release notes. Given the sheer amount of work on twinBASIC already, I'd expect such breaks to be edge-cases rather than anything fundamental.

Perhaps such breaking changes should be held until the next major version of twinBASIC is released, unless they are critical?

**Comment by bclothier** [comment]
Well, if there's a bug that impedes VB6 compatibility it's not achieving the goal and thus not really a "breaking change", no? 

But I take your point about the tB code breaking due to bugs like such but I think that as long there's a corresponding inspection in the problem pane explaining that this code is no longer supported with a way to fix it, that should help alleviate any issues arising from tB codebases that depends on this bug.

**Comment by mansellan** [comment]
Agree.

.Net Core started as a rewrite of .Net Framework, but later had to accept some legacy to keep everyone on board.

twinBASIC comes from a different place. It's not a rewrite, it"# a successor.

**Comment by mwolfe02** [comment]
> Perhaps such breaking changes should be held until the next major version of twinBASIC is released, unless they are critical?

I'm not sure how much extra work it would be for Wayne, but maybe any post-tB1 VBx incompatibilities that _are also_ tB1 breaking changes should immediately go into a tB2-alpha.  Basically, a fork of tB1 with VBx bug fixes that break compatibility with tB2.  If the goal is to follow [semantic versioning](https://semver.org), that seems one way to do it without blowing through a bunch of major twinBASIC version numbers.
